<div style="display: flex; background-color: RGB(31,42,68);">
<h1 style="margin: auto; padding: 30px; color: white;">ANALYSE DES VENTES DE LA LIBRAIRIE LAPAGE</h1>
</div>

# Analyse des données sourcées de la librairie Lapage à partir des fichiers transactions.csv, products.csv et customers.csv. 




<div style="background-color: #F4F6F7;" >
<h2 style="margin: auto; padding: 20px; color:#263E50; ">PHASE I : ANALYSE EXPLORATOIRE DES DONNEES</h2>
</div>


### Objectif : préparer un dataset unifié (jointure des fichiers) prêt pour les analyses et les visualisations demandées par la direction marketing (Annabelle) et la Bi analyst (Julie) afin de présenter des insights actionnables pour la prise de décision stratégique et opérationnelle au CODIR (pour Sylvain).

<div style="background-color: #2E5EAA; border: 1px solid #2E5EAA;" >
<h3 style="margin: auto; padding: 20px; color: #F4F6F7; "><a id="RNCP37837BC01"></a>1.1 - Configuration de l'environnement</h3>
</div>

<div style="font-size: 12px; color: #666; margin-top: 10px;">
<strong>Bloc compétence :</strong> <span style="background-color: #E8F4F8; padding: 2px 6px; border-radius: 3px;">RNCP37837BC01</span> - Structurer et gérer la base de données
</div>

In [1]:
# Import des bibliothèques
import pandas as pd # pour la manipulation de données
import numpy as np # pour les calculs numériques
import matplotlib.pyplot as plt # pour la visualisation
import seaborn as sns # pour la visualisation avancée
from datetime import date # pour la manipulation de dates

# Configuration de l'affichage
pd.set_option('display.max_columns', None)  # Afficher toutes les colonnes
pd.set_option('display.max_rows', 100)      # Limiter à 100 lignes
pd.set_option('display.width', None)        # Pas de limite de largeur

# Style des graphiques
plt.style.use('seaborn-v0_8-whitegrid') # Style de base pour les graphiques
plt.rcParams['figure.figsize'] = (12, 6) # Taille par défaut des figures
plt.rcParams['font.size'] = 12 # Taille de police par défaut

sns.set_theme(style="whitegrid") # Thème de Seaborn pour les graphiques

plt.rcParams.update({
    "figure.facecolor": "white",
    "axes.facecolor": "#F4F6F7",
    "axes.edgecolor": "#1F2A44",
    "axes.labelcolor": "#2C3E50",
    "axes.titleweight": "bold",
    "axes.titlesize": 14,
    "axes.titlecolor": "#1F2A44",
    "xtick.color": "#2C3E50",
    "ytick.color": "#2C3E50",
    "grid.color": "#D5D8DC",
    "font.size": 11
})

sns.set_palette([
    "#2E5EAA",   # bleu principal
    "#C2188B",   # accent
    "#2E8B57",   # validation
    "#7F8C8D",   # neutre
    "#E67E22",   # avertissement
])
print("✅ Bibliothèques chargées avec succès !")

✅ Bibliothèques chargées avec succès !


<div style="background-color: #2E5EAA; border: 1px solid #2E5EAA;" >
<h3 style="margin: auto; padding: 20px; color: #F4F6F7; "><a id="RNCP37837BC02-COLLECT"></a>1.2 - Chargement des fichiers</h3>
</div>

<div style="font-size: 12px; color: #666; margin-top: 10px;">
<strong>Bloc compétence :</strong> <span style="background-color: #E8F4F8; padding: 2px 6px; border-radius: 3px;">RNCP37837BC02</span> - Identifier, collecter et analyser les données
</div>

In [2]:
# Définir le chemin vers les données
DATA_PATH = "../../data/"

# Chargement des 3 fichiers
customers = pd.read_csv(DATA_PATH + "customers.csv", sep=";")
products = pd.read_csv(DATA_PATH + "products.csv", sep=";")
transactions = pd.read_csv(DATA_PATH + "transactions.csv", sep=";")

print("✅ Fichiers chargés avec succès !")
print(f"   - customers    : {len(customers):,} lignes")
print(f"   - products     : {len(products):,} lignes")
print(f"   - transactions : {len(transactions):,} lignes")

✅ Fichiers chargés avec succès !
   - customers    : 8,621 lignes
   - products     : 3,286 lignes
   - transactions : 1,048,575 lignes


C:\Users\feria\AppData\Local\Temp\ipykernel_32320\3126890582.py:7: DtypeWarning: Columns (0,1,2,3) have mixed types. Specify dtype option on import or set low_memory=False.
  transactions = pd.read_csv(DATA_PATH + "transactions.csv", sep=";")


<div style="background-color: #2E5EAA; border: 1px solid #2E5EAA;" >
<h3 style="margin: auto; padding: 20px; color: #F4F6F7; "><a id="RNCP37837BC02-ANALYZE"></a>1.3 - Inspection des fichiers</h3>
</div>

<div style="font-size: 12px; color: #666; margin-top: 10px;">
<strong>Bloc compétence :</strong> <span style="background-color: #E8F4F8; padding: 2px 6px; border-radius: 3px;">RNCP37837BC02</span> - Explorer et pré-traiter les données
</div>

In [3]:
# Inspection de base des fichiers grâce à la fonction inspecter_dataframes()
import os
import sys
sys.path.append(os.path.abspath(os.path.join(os.getcwd(), "../..", "src")))

from data_manager import inspecter_dataframe


inspecter_dataframe(customers, nom="customers")
inspecter_dataframe(products, nom="products")
inspecter_dataframe(transactions, nom="transactions")



📊 RAPPORT D'INSPECTION : CUSTOMERS

📐 DIMENSIONS
   - Nombre de lignes   : 8,621
   - Nombre de colonnes : 3

📋 COLONNES ET TYPES
----------------------------------------
   client_id                 → object
   sex                       → object
   birth                     → int64

❓ VALEURS MANQUANTES (NA)
----------------------------------------
   ✅ Aucune valeur manquante !

🔑 RECHERCHE DE CLÉ UNIQUE (IDENTIFIANT)
----------------------------------------
   ✅ 'client_id' est une clé unique (8,621 valeurs uniques)

📈 STATISTIQUES DESCRIPTIVES
----------------------------------------

   Colonnes numériques :


,birth
count,8621.00
mean,1978.28
std,16.92
min,1929.00
25%,1966.00
50%,1979.00
75%,1992.00
max,2004.00



   Colonnes catégorielles :


,client_id,sex
count,8621,8621
unique,8621,2
top,c_4410,f
freq,1,4490



Fin du rapport pour customers


📊 RAPPORT D'INSPECTION : PRODUCTS

📐 DIMENSIONS
   - Nombre de lignes   : 3,286
   - Nombre de colonnes : 3

📋 COLONNES ET TYPES
----------------------------------------
   id_prod                   → object
   price                     → float64
   categ                     → int64

❓ VALEURS MANQUANTES (NA)
----------------------------------------
   ✅ Aucune valeur manquante !

🔑 RECHERCHE DE CLÉ UNIQUE (IDENTIFIANT)
----------------------------------------
   ✅ 'id_prod' est une clé unique (3,286 valeurs uniques)

📈 STATISTIQUES DESCRIPTIVES
----------------------------------------

   Colonnes numériques :


,price,categ
count,3286.00,3286.00
mean,21.86,0.37
std,29.85,0.62
min,0.62,0.00
25%,6.99,0.00
50%,13.08,0.00
75%,22.99,1.00
max,300.00,2.00



   Colonnes catégorielles :


,id_prod
count,3286
unique,3286
top,0_1421
freq,1



Fin du rapport pour products


📊 RAPPORT D'INSPECTION : TRANSACTIONS

📐 DIMENSIONS
   - Nombre de lignes   : 1,048,575
   - Nombre de colonnes : 4

📋 COLONNES ET TYPES
----------------------------------------
   id_prod                   → object
   date                      → object
   session_id                → object
   client_id                 → object

❓ VALEURS MANQUANTES (NA)
----------------------------------------
   ⚠️  Total de valeurs manquantes : 1,444,164
   - id_prod                   : 361,041 (34.4%)
   - date                      : 361,041 (34.4%)
   - session_id                : 361,041 (34.4%)
   - client_id                 : 361,041 (34.4%)

🔑 RECHERCHE DE CLÉ UNIQUE (IDENTIFIANT)
----------------------------------------
   ❌ Aucune colonne n'est une clé unique
   💡 Les colonnes avec le plus de valeurs uniques :
      - id_prod: 3,265 valeurs uniques
      - date: 687,419 valeurs uniques
      - session_id: 345,505 valeurs uniques

📈 STATISTIQUES DESCRIPTIVES
--

,id_prod,date,session_id,client_id
count,687534,687534,687534,687534
unique,3265,687419,345505,8600
top,1_369,2021-10-13 17:05:59.572162,s_118668,c_1609
freq,2340,2,14,25586



Fin du rapport pour transactions



<div style="background-color: #2E5EAA; border: 1px solid #2E5EAA;" >
<h3 style="margin: auto; padding: 20px; color: #F4F6F7; "><a id="RNCP37837BC02-CLEAN"></a>1.4 - Nettoyage des données</h3>
</div>

<div style="font-size: 12px; color: #666; margin-top: 10px;">
<strong>Blocs compétences :</strong> <span style="background-color: #E8F4F8; padding: 2px 6px; border-radius: 3px;">RNCP37837BC02</span> - Traiter les données manquantes, vérifier la cohérence
</div>

#### Objectif : préparer les données pour l'analyse en corrigeant les erreurs, en traitant les valeurs manquantes et en formatant les données de manière appropriée.

##### 1.4.1 Transactions :   


In [4]:
# Combien de lignes complètement vides ?
print("Lignes entièrement vides :", transactions.isna().all(axis=1).sum())

# Combien de lignes avec au moins 1 NaN ?
print("Lignes avec au moins 1 NaN :", transactions.isna().any(axis=1).sum())

# Détail par colonne
print("\nNaN par colonne :\n", transactions.isna().sum())

Lignes entièrement vides : 361041
Lignes avec au moins 1 NaN : 361041

NaN par colonne :
 id_prod       361041
date          361041
session_id    361041
client_id     361041
dtype: int64


 Conclusion : Des lignes vides dans la table de faits Transactions seraient à l'origine de problèmes dans les étapes suivantes. ==> Décision : Nous allons les supprimer. Il faut également convertir les champs de date en format datetime pour faciliter les analyses temporelles.

In [5]:
# Suppression des lignes entièrement vides
transactions.dropna(how='all', inplace=True)

# Conversion de la date en datetime
transactions['date'] = pd.to_datetime(transactions['date'], errors='coerce')

print("Lignes après nettoyage :", transactions.shape[0])
print("NaN restants :\n", transactions.isna().sum())

Lignes après nettoyage : 687534
NaN restants :
 id_prod       0
date          0
session_id    0
client_id     0
dtype: int64


##### 1.4.2 customers :   


In [6]:
print("NaN customers :\n", customers.isna().sum())
print("Duplicated customers :\n", customers.duplicated().sum())
print("\nTypes :\n", customers.dtypes)
print("\nAperçu :\n", customers.head())

NaN customers :
 client_id    0
sex          0
birth        0
dtype: int64
Duplicated customers :
 0

Types :
 client_id    object
sex          object
birth         int64
dtype: object

Aperçu :
   client_id sex  birth
0    c_4410   f   1967
1    c_7839   f   1975
2    c_1699   f   1984
3    c_5961   f   1962
4    c_5320   m   1943


Diagnostic : aucune valeur manquante , la clé d'identification client est unique, pas de doublons. Nous allons conserver la table customers telle quelle.

##### 1.4.3 products :   


In [7]:
print("NaN products :\n", products.isna().sum())
print("Duplicated products :\n", products.duplicated().sum())
print("\nTypes :\n", products.dtypes)
print("\nAperçu :\n", products.head())

NaN products :
 id_prod    0
price      0
categ      0
dtype: int64
Duplicated products :
 0

Types :
 id_prod     object
price      float64
categ        int64
dtype: object

Aperçu :
   id_prod  price  categ
0  0_1421  19.99      0
1  0_1368   5.13      0
2   0_731  17.99      0
3   1_587   4.99      1
4  0_1507   3.99      0


Diagnostic : aucune valeur manquante , la clé d'identification produit est unique, pas de doublons. Nous allons conserver la table products telle quelle avec id_prod pour la clé de jointure.

In [8]:
#vérification des doublons
print(f"Products     : {products.duplicated().sum()} doublon(s)")
print(f"Customers    : {customers.duplicated().sum()} doublon(s)")
print(f"Transactions : {transactions.duplicated().sum()} doublon(s)")

Products     : 0 doublon(s)
Customers    : 0 doublon(s)
Transactions : 0 doublon(s)


Conclusion: Après suppression des  lignes vides dans la table de faits Transactions, il n'y a plus de valeurs manquantes. De plus, il n'y a pas de doublons dans les tables de dimensions. Nous allons donc conserver les données telles quelles pour les étapes suivantes de jointures afin d'obtenir un dataset unifié prêt pour les analyses et les visualisations. Taille de l'échantillon : 687534 lignes, 100% des données intiales. 

<div style="background-color: #2E5EAA; border: 1px solid #2E5EAA;" >
<h3 style="margin: auto; padding: 20px; color: #F4F6F7; "><a id="RNCP37837BC02-UNIVARIATE"></a>1.5 - Statistiques descriptives</h3>
</div>

<div style="font-size: 12px; color: #666; margin-top: 10px;">
<strong>Bloc compétence :</strong> <span style="background-color: #E8F4F8; padding: 2px 6px; border-radius: 3px;">RNCP37837BC02</span> - Analyse univariée/multivariée
</div>

Objectif : mieux comprendre les données, identifier les tendances, les anomalies et les relations entre les différentes variables.

In [9]:
print("=== PRODUCTS ===")
print(products.describe(include='all').to_string())



=== PRODUCTS ===
       id_prod        price        categ
count     3286  3286.000000  3286.000000
unique    3286          NaN          NaN
top     0_1421          NaN          NaN
freq         1          NaN          NaN
mean       NaN    21.863597     0.370359
std        NaN    29.849786     0.615446
min        NaN     0.620000     0.000000
25%        NaN     6.990000     0.000000
50%        NaN    13.075000     0.000000
75%        NaN    22.990000     1.000000
max        NaN   300.000000     2.000000


In [10]:
print("\n=== CUSTOMERS ===")
print(customers.describe(include='all').to_string())




=== CUSTOMERS ===
       client_id   sex        birth
count       8621  8621  8621.000000
unique      8621     2          NaN
top       c_4410     f          NaN
freq           1  4490          NaN
mean         NaN   NaN  1978.275606
std          NaN   NaN    16.917958
min          NaN   NaN  1929.000000
25%          NaN   NaN  1966.000000
50%          NaN   NaN  1979.000000
75%          NaN   NaN  1992.000000
max          NaN   NaN  2004.000000


In [11]:
print("\n=== TRANSACTIONS ===")
print(transactions.describe(include='all').to_string())


=== TRANSACTIONS ===
       id_prod                           date session_id client_id
count   687534                         687534     687534    687534
unique    3265                            NaN     345505      8600
top      1_369                            NaN   s_118668    c_1609
freq      2340                            NaN         14     25586
mean       NaN  2022-03-01 21:24:00.618519296        NaN       NaN
min        NaN     2021-03-01 00:01:07.843138        NaN       NaN
25%        NaN  2021-09-10 10:35:20.642323456        NaN       NaN
50%        NaN  2022-02-27 06:50:25.400120064        NaN       NaN
75%        NaN  2022-08-28 22:16:49.841665536        NaN       NaN
max        NaN     2023-02-28 23:58:30.792755        NaN       NaN


<div style="background-color: #2E5EAA; border: 1px solid #2E5EAA;" >
<h3 style="margin: auto; padding: 20px; color: #F4F6F7; "><a id="RNCP37837BC02-COHERENCE"></a>1.6 - Recherche d'anomalies dans le catalogue produits et le référentiel Clients par rapport aux Transactions</h3>
</div>

<div style="font-size: 12px; color: #666; margin-top: 10px;">
<strong>Bloc compétence :</strong> <span style="background-color: #E8F4F8; padding: 2px 6px; border-radius: 3px;">RNCP37837BC02</span> - Vérifier la cohérence des données structurées
</div>

##### 1.6.1 products :   
- Vérifier que les prix sont cohérents (pas de prix négatifs ou aberrants).
- Vérifier que les catégories sont correctement attribuées ( pas de valeurs hors de 0,1,2)
- Vérifier que les identifiants de produits sont uniques et cohérents avec les transactions.
- Vérifier la cohérence des données de produits avec les transactions (ex. pas de transactions pour des produits inexistants).

In [12]:
#Vérification des prix négatifs ou aberrants
print("Prix négatifs ou aberrants :\n", products[products['price'] < 0])

Prix négatifs ou aberrants :
 Empty DataFrame
Columns: [id_prod, price, categ]
Index: []


In [13]:
#Vérification que les catégories sont dans les valeurs attendues (0,1,2)
print("Catégories hors de 0,1,2 :\n", products[~products['categ'].isin([0, 1, 2])])

Catégories hors de 0,1,2 :
 Empty DataFrame
Columns: [id_prod, price, categ]
Index: []


In [14]:
#Vérification que les identifiants de produits sont uniques et cohérents avec les transactions
print("IDs de produits dans transactions mais pas dans products :\n", transactions[~transactions['id_prod'].isin(products['id_prod'])]['id_prod'].unique())

IDs de produits dans transactions mais pas dans products :
 []


In [15]:
# Produits jamais vendus
produits_vendus = transactions['id_prod'].unique()
produits_jamais_vendus = products[~products['id_prod'].isin(produits_vendus)]
print(f"Produits jamais vendus : {len(produits_jamais_vendus)} sur {len(products)}")
print(produits_jamais_vendus)

Produits jamais vendus : 21 sur 3286
     id_prod   price  categ
184   0_1016   35.06      0
279   0_1780    1.67      0
736   0_1062   20.08      0
793   0_1119    2.99      0
810   0_1014    1.15      0
845      1_0   31.82      1
1030  0_1318   20.92      0
1138  0_1800   22.05      0
1346  0_1645    2.99      0
1504   0_322    2.99      0
1529  0_1620    0.80      0
1542  0_1025   24.99      0
1708    2_87  220.99      2
1862   1_394   39.73      1
1945    2_72  141.32      2
2214   0_310    1.94      0
2407  0_1624   24.50      0
2524    2_86  132.36      2
2689   0_299   22.99      0
3030   0_510   23.66      0
3095  0_2308   20.28      0


##### 1.6.2 customers :  
- Vérifier que les âges des clients sont cohérents (pas de clients nés avant 1920 ou après 2005).
- Vérifier que les identifiants de clients sont uniques et cohérents avec les transactions.
- Vérifier la cohérence des données de clients avec les transactions (ex. pas de transactions pour des clients inexistants).

In [16]:
#Vérification des âges des clients
print("Âges incohérents :\n", customers[(customers['birth'] < 1920) | (customers['birth'] > 2005)])

Âges incohérents :
 Empty DataFrame
Columns: [client_id, sex, birth]
Index: []


In [17]:
#vérifier que les identifiants sont uniques et cohérents avec les transactions 
print("IDs de clients dans transactions mais pas dans customers :\n", transactions[~transactions['client_id'].isin(customers['client_id'])]['client_id'].unique())

IDs de clients dans transactions mais pas dans customers :
 []


In [18]:
# Clients sans transaction
clients_avec_achat = transactions['client_id'].unique()
clients_sans_achat = customers[~customers['client_id'].isin(clients_avec_achat)]
print(f"\nClients sans transaction : {len(clients_sans_achat)} sur {len(customers)}")
print(clients_sans_achat)


Clients sans transaction : 21 sur 8621
     client_id sex  birth
801     c_8253   f   2001
2483    c_3789   f   1997
2734    c_4406   f   1998
2768    c_2706   f   1967
2850    c_3443   m   1959
3178    c_4447   m   1956
3189    c_3017   f   1992
3333    c_4086   f   1992
3720    c_6930   m   2004
3820    c_4358   m   1999
4723    c_8381   f   1965
4790    c_1223   m   1963
6111    c_6862   f   2002
6207    c_5245   f   2004
6337    c_5223   m   2003
6470    c_6735   m   2004
6584     c_862   f   1956
6827    c_7584   f   1960
7789      c_90   m   2001
7818     c_587   m   1993
8534    c_3526   m   1956


In [19]:
# Ajouter une colonne statut_actif dans customers (version simple)
if 'customers' not in globals() or 'transactions' not in globals():
    print("Veuillez d'abord exécuter les cellules de chargement des fichiers (customers + transactions).")
else:
    clients_avec_achat = transactions['client_id'].dropna().unique()
    customers['statut_actif'] = customers['client_id'].isin(clients_avec_achat)
    customers['statut_actif'] = customers['statut_actif'].map({True: 'Actif', False: 'Inactif'})

    print(customers['statut_actif'].value_counts())
    display(customers[['client_id', 'statut_actif']].head(10))

statut_actif
Actif      8600
Inactif      21
Name: count, dtype: int64


,client_id,statut_actif
0,c_4410,Actif
1,c_7839,Actif
2,c_1699,Actif
3,c_5961,Actif
4,c_5320,Actif
5,c_415,Actif
6,c_285,Actif
7,c_160,Actif
8,c_6446,Actif
9,c_6866,Actif


In [20]:
# Segmenter les clients en BtoB / BtoC avec un seuil de CA
seuil_ca_btob = 100000

# Ici, transactions ne contient pas encore 'price' dans cette section du notebook.
# On recalcule donc un CA client via une jointure temporaire avec products.
transactions_ca = transactions.merge(
    products[['id_prod', 'price']],
    on='id_prod',
    how='left'
 )

# Sécurité: si des prix sont manquants après jointure, on les met à 0
transactions_ca['price'] = transactions_ca['price'].fillna(0)

# CA total par client
ca_client = transactions_ca.groupby('client_id', as_index=False)['price'].sum()
ca_client.columns = ['client_id', 'ca_total']

# Segment selon le seuil
ca_client['segment_client'] = np.where(
    ca_client['ca_total'] > seuil_ca_btob,
    'BtoB',
    'BtoC'
 )

# Ajouter le segment dans customers
customers = customers.merge(ca_client[['client_id', 'ca_total', 'segment_client']], on='client_id', how='left')
customers['segment_client'] = customers['segment_client'].fillna('Sans achat')
customers['ca_total'] = customers['ca_total'].fillna(0)

print(f"Seuil BtoB : {seuil_ca_btob:,.0f} €")
print("\nRépartition des segments :")
print(customers['segment_client'].value_counts())

print("\nAperçu clients BtoB :")
display(customers[customers['segment_client'] == 'BtoB'][['client_id', 'ca_total', 'segment_client']].head(10))

Seuil BtoB : 100,000 €

Répartition des segments :
segment_client
BtoC          8596
Sans achat      21
BtoB             4
Name: count, dtype: int64

Aperçu clients BtoB :


,client_id,ca_total,segment_client
1378,c_4958,290227.03,BtoB
1911,c_6714,153918.60,BtoB
3641,c_1609,326039.89,BtoB
8087,c_3454,114110.57,BtoB


 Contrôle qualité des données  : les données sont propres, cohérentes et prêtes pour les étapes suivantes de jointures et d'analyses. Nous allons documenter les étapes de nettoyage et les décisions prises pour assurer la traçabilité du processus.

# Diagnostic des donnees

Voici la synthese des points cles issus des controles qualite et des statistiques descriptives.

---

## PRODUCTS

### Identifiant produit
- 3286 valeurs uniques pour 3286 lignes
- Aucun doublon detecte

### Prix (price)
- Min : 0.62
- Max : 300
- Moyenne : 21.86
- Mediane (50%) : 13.08
- Ecart-type : 29.85

Interpretation :
- Forte dispersion des prix
- Distribution asymetrique probable
- Presence possible de produits premium

Point de vigilance :
- Verifier les outliers de prix (tres eleves / tres faibles)

### Categorie (categ)
- 3 categories observees : 0, 1, 2
- Majorite en categorie 0

A approfondir :
- Repartition par categorie
- Contribution de chaque categorie au chiffre d'affaires

---

## CUSTOMERS

### Identifiant client
- 8621 valeurs uniques pour 8621 lignes
- Aucun doublon detecte

### Sexe (sex)
- 2 modalites
- Modalite dominante : f (4490 occurrences)

### Annee de naissance (birth)
- De 1929 a 2004
- Moyenne : 1978

Points d'attention :
- Cohérence des ages
- Segmentation par classes d'age

### Activite client sur la periode
- 21 clients sans transaction identifies
- Colonne statut_actif ajoutee : Actif / Inactif
- Les clients sans achat sont conserves dans le referentiel clients

### Segmentation BtoB / BtoC
- Segmentation realisee a partir du CA client cumule
- Regle : BtoB si ca_total > 100000 EUR, sinon BtoC
- Les clients sans achat sont classes en Sans achat

---

## TRANSACTIONS

### Qualite des donnees
- Lignes completement vides supprimees
- Conversion de la date en datetime
- Donnees pretes pour jointures et analyses

### Temporalite
- Periode couverte : 2021 a 2023

A analyser ensuite :
- Saisonnalite
- Ruptures de tendance
- Evolution mensuelle du CA, des clients actifs et du volume de transactions

---

## Decision methodologique

La suite de l'analyse reposera sur :
- Statistiques descriptives completes
- Jointure des tables transactions, products, customers
- Feature engineering (age, mois, annee, labels de categorie)
- Analyses business par segment (BtoB/BtoC, actifs/inactifs)

Objectif : produire un dataset unifie fiable et interpretable pour les analyses statistiques et les visualisations de la phase suivante.

In [21]:
transactions.to_csv(DATA_PATH + "processed/1-nettoyage_transactions.csv", index=False)

print("✅ Données sauvegardées dans 'data/processed/1-nettoyage_transactions.csv'")

✅ Données sauvegardées dans 'data/processed/1-nettoyage_transactions.csv'


<div style="background-color: #2E5EAA; border: 1px solid #2E5EAA;" >
<h3 style="margin: auto; padding: 20px; color: #F4F6F7; ">1.7 - Jointure des tables</h3>
</div>

Les 3 fichiers de données possèdent des clés communes qui permettent de les relier entre eux en respectant les règles d'intégrité référentielle. A ce stade, nous souhaitons conserver toutes les lignes de la table des transactions et ne conserver que les lignes correspondantes dans les tables des produits et des clients. Nous allons donc effectuer des jointures de type "Left join" en utilisant les clés suivantes : client_id pour relier transactions et customers, et id_prod pour relier transactions et products.

In [22]:
# Jointure transactions + products
df = transactions.merge(products, on='id_prod', how='left')


print("Shape après jointure :", df.shape)
print("\nAperçu :\n", df.head())
print("\nNaN après jointure :\n", df.isna().sum())

Shape après jointure : (687534, 6)

Aperçu :
   id_prod                       date session_id client_id  price  categ
0  0_1259 2021-03-01 00:01:07.843138        s_1     c_329  11.99      0
1  0_1390 2021-03-01 00:02:26.047414        s_2     c_664  19.37      0
2  0_1352 2021-03-01 00:02:38.311413        s_3     c_580   4.50      0
3  0_1458 2021-03-01 00:04:54.559692        s_4    c_7912   6.55      0
4  0_1358 2021-03-01 00:05:18.801198        s_5    c_2033  16.49      0

NaN après jointure :
 id_prod       0
date          0
session_id    0
client_id     0
price         0
categ         0
dtype: int64


In [23]:
# Jointure avec customers
df = df.merge(customers, on='client_id', how='left')

print("Shape après jointure customers :", df.shape)
print("\nAperçu :\n", df.head())
print("\nNaN après jointure :\n", df.isna().sum())

Shape après jointure customers : (687534, 11)

Aperçu :
   id_prod                       date session_id client_id  price  categ sex  \
0  0_1259 2021-03-01 00:01:07.843138        s_1     c_329  11.99      0   f   
1  0_1390 2021-03-01 00:02:26.047414        s_2     c_664  19.37      0   m   
2  0_1352 2021-03-01 00:02:38.311413        s_3     c_580   4.50      0   m   
3  0_1458 2021-03-01 00:04:54.559692        s_4    c_7912   6.55      0   f   
4  0_1358 2021-03-01 00:05:18.801198        s_5    c_2033  16.49      0   f   

   birth statut_actif  ca_total segment_client  
0   1967        Actif   1132.94           BtoC  
1   1960        Actif   2314.74           BtoC  
2   1988        Actif   4331.69           BtoC  
3   1989        Actif   3022.26           BtoC  
4   1956        Actif   1285.06           BtoC  

NaN après jointure :
 id_prod           0
date              0
session_id        0
client_id         0
price             0
categ             0
sex               0
birth      

<div style="background-color: #2E5EAA; border: 1px solid #2E5EAA;" >
<h3 style="margin: auto; padding: 20px; color: #F4F6F7; ">1.8 - Feature Engineering</h3>
</div>

- Calcul de l'âge des clients à partir de leur date de naissance pour faciliter les analyses démographiques 
- calcul des données Mois , Année à partir de la date de transaction pour faciliter les analyses temporelles et saisonnières.
- Création d'un libellé de catégorie pour faciliter les analyses de segmentation produits

In [24]:
# Calcul de l'âge
df['date'] = pd.to_datetime(df['date'], errors='coerce')
df['birth'] = pd.to_numeric(df['birth'], errors='coerce')
df['age'] = df['date'].dt.year - df['birth']

print("Shape après jointure :", df.shape)
print("\nAperçu :\n", df.head())
print("\nNaN après jointure :\n", df.isna().sum())

Shape après jointure : (687534, 12)

Aperçu :
   id_prod                       date session_id client_id  price  categ sex  \
0  0_1259 2021-03-01 00:01:07.843138        s_1     c_329  11.99      0   f   
1  0_1390 2021-03-01 00:02:26.047414        s_2     c_664  19.37      0   m   
2  0_1352 2021-03-01 00:02:38.311413        s_3     c_580   4.50      0   m   
3  0_1458 2021-03-01 00:04:54.559692        s_4    c_7912   6.55      0   f   
4  0_1358 2021-03-01 00:05:18.801198        s_5    c_2033  16.49      0   f   

   birth statut_actif  ca_total segment_client  age  
0   1967        Actif   1132.94           BtoC   54  
1   1960        Actif   2314.74           BtoC   61  
2   1988        Actif   4331.69           BtoC   33  
3   1989        Actif   3022.26           BtoC   32  
4   1956        Actif   1285.06           BtoC   65  

NaN après jointure :
 id_prod           0
date              0
session_id        0
client_id         0
price             0
categ             0
sex        

In [25]:
# Extraction mois et année
df['mois'] = df['date'].dt.to_period('M')
df['annee'] = df['date'].dt.year

print("Aperçu des nouvelles colonnes :\n", df[['date', 'birth', 'age', 'mois', 'annee']].head())

Aperçu des nouvelles colonnes :
                         date  birth  age     mois  annee
0 2021-03-01 00:01:07.843138   1967   54  2021-03   2021
1 2021-03-01 00:02:26.047414   1960   61  2021-03   2021
2 2021-03-01 00:02:38.311413   1988   33  2021-03   2021
3 2021-03-01 00:04:54.559692   1989   32  2021-03   2021
4 2021-03-01 00:05:18.801198   1956   65  2021-03   2021


In [26]:
# Labels lisibles pour les catégories
categ_labels = {0: 'Catégorie 0', 1: 'Catégorie 1', 2: 'Catégorie 2'}
df['categ_label'] = df['categ'].map(categ_labels)
print("Aperçu des nouvelles colonnes :\n", df[['date', 'birth', 'age', 'mois', 'annee','categ_label']].head())

Aperçu des nouvelles colonnes :
                         date  birth  age     mois  annee  categ_label
0 2021-03-01 00:01:07.843138   1967   54  2021-03   2021  Catégorie 0
1 2021-03-01 00:02:26.047414   1960   61  2021-03   2021  Catégorie 0
2 2021-03-01 00:02:38.311413   1988   33  2021-03   2021  Catégorie 0
3 2021-03-01 00:04:54.559692   1989   32  2021-03   2021  Catégorie 0
4 2021-03-01 00:05:18.801198   1956   65  2021-03   2021  Catégorie 0


In [27]:
#inspection du dataset final
inspecter_dataframe(df, nom='Dataset final')


📊 RAPPORT D'INSPECTION : DATASET FINAL

📐 DIMENSIONS
   - Nombre de lignes   : 687,534
   - Nombre de colonnes : 15

📋 COLONNES ET TYPES
----------------------------------------
   id_prod                   → object
   date                      → datetime64[ns]
   session_id                → object
   client_id                 → object
   price                     → float64
   categ                     → int64
   sex                       → object
   birth                     → int64
   statut_actif              → object
   ca_total                  → float64
   segment_client            → object
   age                       → int64
   mois                      → period[M]
   annee                     → int32
   categ_label               → object

❓ VALEURS MANQUANTES (NA)
----------------------------------------
   ✅ Aucune valeur manquante !

🔑 RECHERCHE DE CLÉ UNIQUE (IDENTIFIANT)
----------------------------------------
   ❌ Aucune colonne n'est une clé unique
   💡 Les colonnes av

,price,categ,birth,ca_total,age,annee
count,687534.00,687534.00,687534.00,687534.00,687534.00,687534.00
mean,17.49,0.45,1977.82,19415.15,43.84,2021.66
std,18.24,0.59,13.61,68333.11,13.62,0.62
min,0.62,0.00,1929.00,6.31,17.00,2021.00
25%,8.99,0.00,1970.00,1173.31,35.00,2021.00
50%,13.99,0.00,1980.00,1976.09,42.00,2022.00
75%,19.08,1.00,1987.00,3067.13,51.00,2022.00
max,300.00,2.00,2004.00,326039.89,94.00,2023.00



   Colonnes catégorielles :


,id_prod,session_id,client_id,sex,statut_actif,segment_client,categ_label
count,687534,687534,687534,687534,687534,687534,687534
unique,3265,345505,8600,2,1,2,3
top,1_369,s_118668,c_1609,m,Actif,BtoC,Catégorie 0
freq,2340,14,25586,344841,687534,640734,415459



Fin du rapport pour Dataset final



In [28]:
# sauvegarde du dataframe final après nettoyage, jointure et feature engineering
df.to_csv(DATA_PATH + "processed/2-transactions_enrichies.csv", index=False)

<div style="background-color: #F4F6F7;" >
<h2 style="margin: auto; padding: 20px; color:#263E50; ">FIN  ANALYSE EXPLORATOIRE DES DONNEES</h2>
</div>